src/mono/mono/mini/intrinsics.c

```diff
@@ -488,6 +488,124 @@ emit_unsafe_intrinsics (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignatu
+		if (mini_is_gsharedvt_variable_type (tfrom)) {
+			return NULL;
+		}
+
+		MonoType *tto = ctx->method_inst->type_argv [1];
+		if (mini_is_gsharedvt_variable_type (tto)) {
+			return NULL;
+		}
+
+		// The underlying API always throws for reference type inputs, so we
+		// fallback to the managed implementation to let that handling occur
+
+		MonoTypeEnum tfrom_type = tfrom->type;
+		if (MONO_TYPE_IS_REFERENCE (tfrom)) {
+			return NULL;
+		}
+
+		MonoTypeEnum tto_type = tto->type;
+		if (MONO_TYPE_IS_REFERENCE (tto)) {
+			return NULL;
+		}
+
+		// The same applies for when the type sizes do not match, as this will always throw
+		// and so its not an expected case and we can fallback to the managed implementation
+
+		int tfrom_align, tto_align;
+		if (mono_type_size (tfrom, &tfrom_align) != mono_type_size (tto, &tto_align)) {
+			return NULL;
+		}
+
+		// We have several different move opcodes to handle the data depending on the
+		// source and target types, so detect and optimize the most common ones falling
+		// back to what is effectively `ReadUnaligned<TTo>(ref As<TFrom, byte>(ref source))`
+		// for anything that can't be special cased as potentially zero-cost move.
+
+		MonoClass *tfrom_klass = mono_class_from_mono_type_internal (tfrom);
+		MonoClass *tto_klass = mono_class_from_mono_type_internal (tto);
+
+		guint32 opcode = OP_LDADDR;
+		MonoStackType tfrom_stack = STACK_PTR;
+		MonoStackType tto_stack = STACK_OBJ;
+
+		if (tfrom_type == MONO_TYPE_I4) {
+			if (tto_type == MONO_TYPE_R4) {
+				opcode = OP_MOVE_I4_TO_F;
+				tfrom_stack = STACK_I4;
+				tto_stack = STACK_R4;
+			} else if (tto_type == MONO_TYPE_I4) {
+				opcode = OP_MOVE;
+				tfrom_stack = STACK_I4;
+				tto_stack = STACK_I4;
+			}
+		} else if (tfrom_type == MONO_TYPE_I8) {
+			if (tto_type == MONO_TYPE_R8) {
+				opcode = OP_MOVE_I8_TO_F;
+				tfrom_stack = STACK_I8;
+				tto_stack = STACK_R8;
+			} else if (tto_type == MONO_TYPE_I8) {
+				opcode = OP_MOVE;
+				tfrom_stack = STACK_I8;
+				tto_stack = STACK_I8;
+			}
+		} else if (tfrom_type == MONO_TYPE_R4) {
+			if (tto_type == MONO_TYPE_I4) {
+				opcode = OP_MOVE_F_TO_I4;
+				tfrom_stack = STACK_R4;
+				tto_stack = STACK_I4;
+			} else if (tto_type == MONO_TYPE_R4) {
+				opcode = OP_RMOVE;
+				tfrom_stack = STACK_R4;
+				tto_stack = STACK_R4;
+			}
+		} else if (tfrom_type == MONO_TYPE_R8) {
+			if (tto_type == MONO_TYPE_I8) {
+				opcode = OP_MOVE_F_TO_I8;
+				tfrom_stack = STACK_R8;
+				tto_stack = STACK_I8;
+			} else if (tto_type == MONO_TYPE_R8) {
+				opcode = OP_FMOVE;
+				tfrom_stack = STACK_R8;
+				tto_stack = STACK_R8;
+			}
+		} else if (mini_class_is_simd (cfg, tfrom_klass)) {
+			if (mini_class_is_simd (cfg, tto_klass)) {
+				opcode = OP_XMOVE;
+				tfrom_stack = STACK_VTYPE;
+				tto_stack = STACK_VTYPE;
+			}
+		}
+
+		if (opcode == OP_LDADDR) {
+			MonoInst *arg0;
+			NEW_VARLOADA (cfg, arg0, args [0], tfrom);
+			arg0->dreg = mono_alloc_preg (cfg);
```

Comment:
`NEW_VARLOADA` does its own assignment to `arg0->dreg`:

```c
(dest)->dreg = alloc_dreg ((cfg), STACK_MP); \
```

so I don't think this line is needed.

(also the `tfrom` arg is unused by `NEW_VARLOADA`, although it does access `args[0]->klass` - which is the same thing for BitCast)

Reply:
Fixed.